keps/sig-api-machinery/20190927-insecure-backend-proxy.md

```diff
@@ -0,0 +1,201 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+When trying to get logs for a pod, it is possible for a kubelet to have an expired serving certificate.
+If a client chooses, it should be possible to bypass the default behavior of the kube-apiserver and allow the kube-apiserver
+to skip TLS verification of the kubelet to allow gathering logs.  This is safe because the kube-apiserver's credentials
+are always client certificates which cannot be replayed by an evil-kubelet and risk is contained to an evil-kubelet
+returning false log data.  If the user has chosen to accept this risk, we should allow it for the same reason we
+have an option for `--insecure-skip-tls-verify`.
+
+## Motivation
+
+On self-hosted clusters it is possible to end up in a state where a kubelet's serving certificate has expired so a kube-apiserver
+cannot verify the kubelet identity, *but* the kube-apiserver's client certificate is still valid so the kubelet can still
+verify the kube-apiserver.  In this condition, a cluster-admin may need to get pod logs to debug his cluster.
+
+### Goals
+
+1. Allow cluster-admins to get pod logs from kubelets with expired serving certificates.  This will include an API change
+and an addition argument to `kubectl log`
+
+### Non-Goals
+
+1. Allow any bidirectional traffic proxied to kubelets.  This may be a future objective, but is not in scope for the current KEP.
+
+## Proposal
+
+In [PodLogOptions](https://github.com/kubernetes/api/blob/d58b53da08f5430bb0f4e1154a73314e82b5b3aa/core/v1/types.go), 
+add a `InsecureSkipTLSVerifyBackend bool`
+```go
+// PodLogOptions is the query options for a Pod's logs REST call.
+type PodLogOptions struct {
+	// ... existing fields snipped
+	
+	// insecureSkipTLSVerifyBackend indicates that the apiserver should not confirm the validity of the 
+	// serving certificate of the backend it is connecting to.  This will make the HTTPS connection between the apiserver
+	// and the backend insecure. This means that an attacker could intercept the logs request and provide false data.
```

Comment:
I'd recommend clarifying that this means the apiserver can not verify the log data it is receiving came from the real kubelet.

Assuming the kubelet is still verifying the API server's TLS credentials, it does **not** mean the connection to the real kubelet is vulnerable to a man in the middle attack (e.g. an attacker could not intercept the actual log data coming from the real kubelet).

That's important, because we don't want a user action to expose confidential data.

Reply:
Ok, will clarify